adala/runtimes/_litellm.py

```diff
@@ -0,0 +1,537 @@
+            Dict[str, str]: Output record.
+        """
+
+        extra_fields = extra_fields or {}
+        field_schema = field_schema or {}
+
+        options = {}
+        for field, schema in field_schema.items():
+            if schema.get('type') == 'array':
+                options[field] = schema.get('items', {}).get('enum', [])
+
+        output_fields = parse_template(
+            partial_str_format(output_template, **extra_fields),
+            include_texts=True,
+        )
+        system_prompt = instructions_template
+        user_prompt = input_template.format(**record, **extra_fields)
+        messages = [{'role': 'system', 'content': system_prompt}]
+
+        outputs = {}
+        for output_field in output_fields:
+            if output_field['type'] == 'text':
+                if user_prompt is not None:
+                    user_prompt += f"\n{output_field['text']}"
+                else:
+                    user_prompt = output_field['text']
+            elif output_field['type'] == 'var':
+                name = output_field['text']
+                messages.append({'role': 'user', 'content': user_prompt})
+                completion_text = self.execute(messages)
+                if name in options:
+                    completion_text = match_options(
+                        completion_text, options[name]
+                    )
+                outputs[name] = completion_text
+                messages.append(
+                    {'role': 'assistant', 'content': completion_text}
+                )
+                user_prompt = None
+
+        return outputs
+
+
+class AsyncLiteLLMChatRuntime(AsyncRuntime):
+    """
+    Runtime that uses [OpenAI API](https://openai.com/) and chat completion models to perform the skill.
+    It uses async calls to OpenAI API.
+
+    Attributes:
+        openai_model: OpenAI model name.
+        openai_api_key: OpenAI API key. If not provided, will be taken from OPENAI_API_KEY environment variable.
+        base_url: Can point to any implementation of the OpenAI API. Defaults to OpenAI's.
+        max_tokens: Maximum number of tokens to generate. Defaults to 1000.
+        temperature: Temperature for sampling, between 0 and 1. Higher values means the model will take more risks.
+            Try 0.9 for more creative applications, and 0 (argmax sampling) for ones with a well-defined answer.
+            Defaults to 0.0.
+
+        concurrent_clients: Number of concurrent clients to OpenAI API. More clients means more parallel requests, but
+            also more money spent and more chances to hit the rate limit. Defaults to 10.
+    """
+
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True
+    )  # for @computed_field
+
+    model: str
+    api_key: Optional[str] = None
+    base_url: Optional[str] = None
+    max_tokens: Optional[int] = 1000
+    temperature: Optional[float] = 0.0
+    splitter: Optional[str] = None
+    # concurrent_clients: Optional[int] = 10
```

Comment:
left the concurrent clients stuff around in comments until I get around to testing more thoroughly whether we need it or not.
If anyone knows a reason we definitely still need it, or has a great way to test whether we need it, lmk.
Otherwise I'll be doing some kinda broad testing trying to elicit an issue.

Reply:
Without knowing litellm specifics - I figure concurrent clients will help us get results faster right? As long as rate limits are respected I would think 10 clients would be faster than 1, especially for larger projects. Depends on the rate that we are able to send tasks from LSE to Adala of course

Reply:
sure, but do we ever actually create additional clients? Seems we only use this to pass is to httpx, but I could be missing some adala magic here

Reply:
this is legacy from before httpx was used to handle clients and each client was created manually. I don't think there's a benefit to controlling this at the user facing level.